src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -4489,12 +4489,19 @@ void C2_MacroAssembler::count_positives(Register ary1, Register len,
 }
 
 // Compare char[] or byte[] arrays aligned to 4 bytes or substrings.
-void C2_MacroAssembler::arrays_equals(bool is_array_equ, Register ary1, Register ary2,
-                                      Register limit, Register result, Register chr,
-                                      XMMRegister vec1, XMMRegister vec2, bool is_char, KRegister mask) {
+void C2_MacroAssembler::arrays_equals(bool is_array_equ, Register ary1,
+                                      Register ary2, Register limit,
+                                      Register result, Register chr,
+                                      XMMRegister vec1, XMMRegister vec2,
+                                      bool is_char, KRegister mask,
+                                      bool expand_ary2) {
+  // for expand_ary2, limit is the (smaller) size of the second array.
   ShortBranchVerifier sbv(this);
   Label TRUE_LABEL, FALSE_LABEL, DONE, COMPARE_VECTORS, COMPARE_CHAR, COMPARE_BYTE;
 
+  assert((!expand_ary2) || ((expand_ary2) && (UseAVX == 2)),
+         "Expansion only implemented for AVX2");
```

Comment:
BTW, the check in assert could be simplified: `(!expand_ary2 || UseAVX == 2)`

Reply:
I thought this would make the intent explicitly clear.